### DISCUSSION TITLE
What would the process of updating a server-less system look like with planetscale?

### DISCUSSION THREAD

**Discussion by samuelgozi** [planetscale/discussion#433] [Q&A] [answered]
I have a system that consists of multiple server-less (AKA autoscaling) instances that interact with the DB (directly or indirectly).
When we initiate an update, lets say from version x.1 to x.2 some instances of x.1 might still be up and running, and we don't want to shut them down (preferably).

My questions are:
1. Is there a way to communicate at which stage of the update is Planetscale? for example, How do I know that the migration is started/ended? I need synchronise the update of the logical part of the system ( AKA instances).

2. Is it possible to control how long will the "compatibility layer" of "gh-host"(or whatever is in charge of converting between the schemas while the migration is on going), so that even if the migration is done, an old service can still write with the old schema?

4. What is the recommended way to synchronise between the DB changes and the services in a serverless environment?

**Comment by mscoutermarsh** [comment, marked as answer]
😄 great questions.

In general, we recommend only making schema changes that are compatible with both the current and next version of application code.

For example, if we are removing a column in our application. We'll deploy application code that stops using it, then deploy the schema change afterwards.

> Is there a way to communicate at which stage of the update is Planetscale? for example, How do I know that the migration is started/ended? I need synchronise the update of the logical part of the system ( AKA instances).

You could either use the `pscale` CLI or this [API endpoint](https://api-docs.planetscale.com/reference/get-a-deployment).

**Here's an example with pscale:**

```
pscale deploy-request show db-name 400 -f json
{
  "id": "sc3binrr9xwz",
  "number": 400,
  "branch": "maxeng-vbj-opts-schema",
  "into_branch": "main",
  "approved": true,
  "state": "closed",
  "deployment": {
    "state": "complete",
    "deployable": true,
    "queued_at": 1671730206413,
    "started_at": 1671730212418,
    "finished_at": 1671730420295
  },
  "created_at": 1671652082619,
  "updated_at": 1671732241722,
  "closed_at": 1671732241668
}
```
When `state` is `complete`, the tables are completely swapped over to the new schema. Please note, there could be a couple seconds between when Vitess swaps over the schema and when our API reports back the migration is complete.

> Is it possible to control how long will the "compatibility layer" of "gh-host"(or whatever is in charge of converting between the schemas while the migration is on going), so that even if the migration is done, an old service can still write with the old schema?

You could disable "auto cutover" and this will put you in control of when the tables swap over. The old table will be gone though, and all your writes will move to the new schema version. [More details](https://planetscale.com/blog/gated-deployments-addressing-the-complexity-of-schema-deployments-at-scale). It's not possible for your application to write to different versions of the schema.

> What is the recommended way to synchronise between the DB changes and the services in a serverless environment?

Best practice here is the same as for a non-serverless application. Keeping the application code compatible with both the current and future version of the schema. This is really the only way to protect from errors & downtime.

If you have any specific schema/app change scenarios you'd like to share, I'd be happy to share how I'd appoach them.
